src/nvim/lua/treesitter.c

```diff
@@ -1411,26 +1420,44 @@ static int query_next_capture(lua_State *L)
     push_node(L, capture.node, lua_upvalueindex(2));  // [index, node]
 
     // Now check if we need to run the predicates
+    // a predicate is executed only once within the same match.
     uint32_t n_pred;
     ts_query_predicates_for_pattern(query, match.pattern_index, &n_pred);
 
-    if (n_pred > 0 && (ud->max_match_id < (int)match.id)) {
+    if (n_pred == 0) {
+      // No predicates. Return without the match table as a small perf optimization,
+      // because in the iter_captures() method, match table won't be needed.
+      // [index, node]
+      return 2;
+    }
+
+    // A new match is found. the use of `max_match_id` assumed that `ts_query_cursor_next_capture()`
+    // would enumerate captures in the order that "match"es increases monotonoically.
+    // The `ts_query_cursor_next_capture()` would enumerate captures in the order they are found.
+    // TODO(wookayin): However, this is not guaranteed, and indeed assertion fails.
+    // Therefore, when matches are alternating, the information about if predicate is active
+    // won't be maintained, which can still lead to incorrect behavior of predicates.
+    assert(ud->max_match_id <= (int)match.id);
+    if (ud->max_match_id < (int)match.id) {
```

Comment:
>However, I found that ts_query_cursor_next_capture enumerates captures in the order they are found, so actually matches can be alternating; and indeed this assertion easily fails although not (yet) covered by tests.

Can you provide an example where this fails? I'm not sure I understand in what situation an ordered list of captures would result in alternating matches. The treesitter header you linked says

>iterate over all of the individual *captures* in the order that they appear

So iterating over captures in the "order that they appear" implies that matches are also iterated in sequential order. Am I misunderstanding?

Reply:
That's what I also assumed, but with this assertion included I can consistently crash my built nvim in a real-world use. CI tests are also red. Will have to find a repro and add it to test.

I think the header file also mentions:

> Because multiple patterns can match the same set of nodes, one match may contain captures that appear *before* some of the captures from a previous match

so I guess this is when, for example, there are multiple (even duplicate) patterns matching on the exact same node.

Reply:
I think https://github.com/neovim/neovim/issues/27895 is a small example of matches alternating.

Reply:
I've addressed this issue by moving all the match tracking into Lua in #27921 and then adding memoization (via the match id) in #28033.